model/library/functions.py

```diff
@@ -0,0 +1,93 @@
+"""Collection of functions"""
+from __future__ import annotations
+from typing import TYPE_CHECKING
+from math import e, pi, sqrt
+import plotly.express as px
+import pandas as pd
+from termcolor import colored
+if TYPE_CHECKING:
+    from model.library.objects import Context, AMI
+
+def get_parameters(context:Context) -> str:
+    """get the simulation parameters"""
+    return (f"Functional Acuity = {context.agent_op.functional_acuity} "+
+        f"stream_shift_amount= {context.instrument.stream_status.stream_shift_amount}, "+
+        f"p_stream_shift={context.instrument.stream_status.p_stream_shift}, "+
+        f"beam_shift_amount={context.instrument.beam_status.beam_shift_amount}, "+
+        f"p_crazy_ivan={context.instrument.stream_status.p_crazy_ivan}")
+
+def calculate_roi(ami:AMI) -> str:
+    """Determine the retun of investment data / time"""
+    return (sum(sample.compleated for sample in ami.samples) / len(ami.samples
+    ) if len(ami.samples) > 0 else 0)
+
+def get_line() -> str:
+    """return string line"""
+    return "---------------------------------------------------------------------------------------"
+
+def goal_agenda_plan(context:Context) -> str:
+    "Return out GAP: Goal Agenda Plan"
+    return (f"{get_parameters(context)}\n{context.ami}\nCurrent Time: {context.current_time} "+
+    f"{context.agenda}\n{get_line()}")
+
+def experiment_stats(context:Context) -> str:
+    "Write out statistics of the experiment"
+    return (f"{context.ami}\nCurrent Time: {context.current_time} {context.agenda}\n"+
+    f"{get_line()}\nROI: {calculate_roi(context.ami):.0%}\n{context.agenda.get_timeline()}")
+
+def experiment_is_not_over(context:Context) -> bool:
+    """True: Experiment is not over, False: Experiment is over"""
+    return (context.current_time < context.agenda.experimental_time
+    ) and (len(context.agenda.event_timeline) != len(context.ami.samples)
+    ) if (len(context.agenda.event_timeline) != 0
+    ) else True
+
+def get_current_datapoints(context:Context) -> str:
+    """Gets the datapoints live"""
+    current_sample = context.ami.get_current_sample(context)
+    return_string = ""
+    for index, data in enumerate(current_sample.data[-60:]):
+        return_string += (colored(f'{data}', 'green'
+        if data.quality >= aquire_data(0.001, current_sample.preformance_quality
+        ) else 'yellow' if data.quality >= aquire_data(0.03, current_sample.preformance_quality
+        ) else 'red') + ("\n" if index == 19 or index == 39 else " "))
+    return return_string
+
+def aquire_data(distance:float, preformance_quality:float) -> float:
```

Comment:
# Important! 
This is there the Gaussian is calculated for the data also one of the three places were the bug might be